File: MIDTERM-LAB-3/mp01.js
/*
 * Student Name   : Tanchingco, John Matthew R.
 * Course Code    : Programming 2
 * Assignment     : Machine Problem 01 - Load Dataset and Display Total Records
 * Date           : 2026-04-08
 * Description    : This Node.js script asks the user for the CSV file path,
 *                  reads and parses the dataset, skips metadata/header rows,
 *                  and displays the total number of valid data records found.
 */

"use strict";

const fs = require("fs");
const readline = require("readline");

// Set up the interface for reading user input from the terminal
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Prompt the user to enter the CSV file path
rl.question("Enter the path to the CSV dataset file: ", (filePath) => {

  console.log("\n===================================================");
  console.log("  MP01 - Load Dataset and Display Total Records");
  console.log("  Student: Tanchingco, John Matthew R.");
  console.log("===================================================\n");

  try {
    // Read the entire file contents as a UTF-8 string
    const fileContents = fs.readFileSync(filePath.trim(), "utf8");

    // Split the file into individual lines for processing
    const lines = fileContents.split("\n");

    // Array to hold the valid data records after skipping metadata
    const records = [];

    // Flag to indicate when we have reached the actual data rows
    let dataStarted = false;

    // Loop through every line and collect valid records
    lines.forEach((line) => {

      // Detect the header row that marks the start of data
      if (line.includes("Candidate") && line.includes("Exam")) {
        dataStarted = true;
        return; // skip the header row itself
      }

      // After the header, add non-empty and non-filler lines to records
      if (dataStarted && line.trim() !== "" && line.trim() !== ",,,,,,,,,,,") {
        records.push(line.trim());
      }
    });

    // Display the results to the user
    console.log("  Dataset loaded successfully.");
    console.log(`  File: ${filePath.trim()}`);
    console.log();
    console.log(`  Total records in dataset: ${records.length}`);

  } catch (err) {
    // Handle file not found or unreadable errors
    console.log("  ERROR: Could not read the file.");
    console.log(`  Details: ${err.message}`);
  }

  console.log("\n===================================================");
  rl.close();
});

File: MIDTERM-LAB-3/MP01.java
/*
 * Student Name   : Tanchingco, John Matthew R.
 * Course Code    : Programming 2
 * Assignment     : Machine Problem 01 - Load Dataset and Display Total Records
 * Date           : 2026-04-08
 * Description    : This program asks the user for the CSV file path, reads the
 *                  dataset using BufferedReader, skips the header/metadata rows,
 *                  counts all valid data records, and displays the total count.
 */

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Scanner;

public class MP01 {

    public static void main(String[] args) {

        // Scanner to read user input from the console
        Scanner scanner = new Scanner(System.in);

        // Prompt the user to enter the path to the CSV dataset
        System.out.print("Enter the path to the CSV dataset file: ");
        String filePath = scanner.nextLine().trim();

        // List to store each valid data record from the CSV
        ArrayList<String> records = new ArrayList<>();

        // Flag to track when we have passed the header/metadata rows
        boolean dataStarted = false;

        System.out.println("\n===================================================");
        System.out.println("  MP01 - Load Dataset and Display Total Records");
        System.out.println("  Student: Tanchingco, John Matthew R.");
        System.out.println("===================================================\n");

        // Open and read the CSV file line by line
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {

            String line;

            // Read each line of the file until end of file
            while ((line = br.readLine()) != null) {

                // The actual data rows begin after the header row containing "Candidate"
                if (line.contains("Candidate") && line.contains("Exam")) {
                    dataStarted = true;
                    continue; // skip the header row itself
                }

                // Once the data section has started, collect non-empty rows
                if (dataStarted && !line.trim().isEmpty() && !line.trim().equals(",,,,,,,,,,,")) {
                    records.add(line);
                }
            }

            // Display the total number of records found in the dataset
            System.out.println("  Dataset loaded successfully.");
            System.out.println("  File: " + filePath);
            System.out.println();
            System.out.println("  Total records in dataset: " + records.size());

        } catch (IOException e) {
            // Handle the case where the file cannot be found or read
            System.out.println("  ERROR: Could not read the file.");
            System.out.println("  Details: " + e.getMessage());
        }

        System.out.println("\n===================================================");

        // Close the scanner to free resources
        scanner.close();
    }
}
